### PR TITLE
MueLu: adding time monitors in region MG

### PR DESCRIPTION
**Pull request by lucbv**
@trilinos/muelu 

## Description
Adding a whole lot of monitors to find were time can be shaved off?
Also fixing a couple of issues in the experimental side of MueLu after default types in Xpetra have been removed.
Finally switching some of the data from `std::vector` to `Teuchos::Array`.

## Motivation and Context
The main goal is to add time monitor for performance diagnostic, it already allowed to identify a target for refactor.
Using `Teuchos::Array` allows us to use parameter list more easily since std::vector cannot be stored natively on parameter list.

## Related Issues

* Closes 
* Blocks 
* Is blocked by 
* Follows 
* Precedes 
* Related to #5656
* Part of 
* Composed of 

## How Has This Been Tested?
Local build and test is successful, the monitors are reporting time appropriately

## Checklist

- [ ] My commit messages mention the appropriate GitHub issue numbers.
- [x] My code follows the code style of the affected package(s).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [code contribution guidelines](../blob/master/CONTRIBUTING.md) for this project.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] No new compiler warnings were introduced.
- [ ] These changes break backwards compatibility.

### CODE REVIEW

File: packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionHierarchy_def.hpp [hunk 1]
```diff
@@ -92,18 +93,18 @@ using Teuchos::ParameterList;
  * This can be achieved by setting \c inverseScaling to \c true.
  */
 template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
-void scaleInterfaceDOFs(std::vector<RCP<Xpetra::Vector<Scalar, LocalOrdinal, GlobalOrdinal, Node> > >& regVec, ///< Vector to be scaled
-    const std::vector<RCP<Xpetra::Vector<Scalar, LocalOrdinal, GlobalOrdinal, Node> > >& scalingFactors, ///< Vector with scaling factors
-    bool inverseScaling ///< Divide by scaling factors (yes/no?)
-    )
+void scaleInterfaceDOFs(Array<RCP<Xpetra::Vector<Scalar, LocalOrdinal, GlobalOrdinal, Node> > >& regVec, ///< Vector to be scaled
```
Comment (mhoemmen) [inline packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionHierarchy_def.hpp]: Regarding the above "def" includes -- why do you need those?  Are you using types for which MueLu does not do ETI?

File: packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionHierarchy_def.hpp [hunk 2]
```diff
@@ -1568,8 +1582,11 @@ computeResidual(std::vector<RCP<Xpetra::Vector<Scalar, LocalOrdinal, GlobalOrdin
     )
 {
 #include "Xpetra_UseShortNames.hpp"
+  using Teuchos::TimeMonitor;
   const int maxRegPerProc = regX.size();
 
+  RCP<TimeMonitor> tm = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("computeResidual: 1 - Rreg = Areg*Xreg")));
```
Comment (mhoemmen) [inline packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionHierarchy_def.hpp]: Did I mention how I feel about handling a scope guard by pointer?  Oh wait I did ;-)

File: packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionSmoothers_def.hpp [hunk 3]
```diff
@@ -105,8 +105,46 @@ void jacobiSetup(RCP<Teuchos::ParameterList> params,
...
+
+    // RCP<Teuchos::FancyOStream> out = Teuchos::fancyOStream(Teuchos::rcpFromRef(std::cout));
+    // regionInterfaceScaling[j]->describe(*out, Teuchos::VERB_EXTREME);
+    // diag[j]->describe(*out, Teuchos::VERB_EXTREME);
+  }
+
+  params->set<Teuchos::Array<RCP<Vector> > >("jacobi: inverse diagonal", diag);
+}
+
+/*! \brief performs Jacobi setup
```
Comment (mhoemmen) [inline packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionSmoothers_def.hpp]: The name kinda says that.  Also, Doxygen documention should go in the declarations header, not in the definitions header.

File: packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionSmoothers_def.hpp [hunk 4]
```diff
@@ -105,8 +105,46 @@ void jacobiSetup(RCP<Teuchos::ParameterList> params,
...
+  Array<RCP<Vector> > regRes(maxRegPerProc);
+  createRegionalVector(regRes, maxRegPerProc, revisedRowMapPerGrp);
+
+  // extract diagonal from region matrices, recover true diagonal values, invert diagonal
+  Teuchos::Array<RCP<Vector> > diag(maxRegPerProc);
+  for (int j = 0; j < maxRegPerProc; j++) {
+    // extract inverse of diagonal from matrix
+    diag[j] = VectorFactory::Build(regionGrpMats[j]->getRowMap(), true);
+    regionGrpMats[j]->getLocalDiagCopy(*diag[j]);
+    diag[j]->elementWiseMultiply(SC_ONE, *regionInterfaceScaling[j], *diag[j], SC_ZERO); // ToDo Does it work to pass in diag[j], but also return into the same variable?
```
Comment (mhoemmen) [inline packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionSmoothers_def.hpp]: Regarding the question in your comment: It works just fine :-)  In general, if there are no cross-loop data dependencies in the obvious `#pragma omp parallel for` implementation, you're fine.

File: packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionHierarchy_def.hpp (hunk 1 above)
Comment (lucbv) [reply packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionHierarchy_def.hpp]: No but this code is not meant to survive all that long so little effort is made to clean it.
A complete redesign will happen before we attempt to move anything out of `Experimental`.
For instance most of the linear algebra related functions will eventually be pulled into classes.

File: packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionHierarchy_def.hpp (hunk 2 above)
Comment (lucbv) [reply packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionHierarchy_def.hpp]: Sorry, I was dumbly copy pasting a pattern used throughout the muelu examples...
I will try to keep this in mind and change the pattern next time I need it : )

File: packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionSmoothers_def.hpp (hunk 3 above)
Comment (lucbv) [reply packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionSmoothers_def.hpp]: Oh there is no declaration header, this is only dirty experimental code...

File: packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionHierarchy_def.hpp (hunk 2 above)
Comment (mhoemmen) [reply packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionHierarchy_def.hpp]: Thanks @lucbv ! :-D  If using a scope guard in the proper way is not convenient, you could always just call `start` and `stop` on the timer.

File: packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionSmoothers_def.hpp [hunk 5]
```diff
@@ -165,7 +203,17 @@ void jacobiIterate(RCP<Teuchos::ParameterList> smootherParams,
...
+    }
+
+    // check for convergence
+    {
+      RCP<Vector> compRes = VectorFactory::Build(mapComp, true);
+      regionalToComposite(regRes, compRes, maxRegPerProc, rowMapPerGrp,
+                          rowImportPerGrp, Xpetra::ADD);
+      typename Teuchos::ScalarTraits<Scalar>::magnitudeType normRes = compRes->norm2();
+
+      if (normRes < 1.0e-12) {return;}
```
Comment (mayrmt) [inline packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionSmoothers_def.hpp]: @lucbv Will this conflict with yesterday's removal of the convergence check from the Jacobi smoother in PR #5701?

**Comment by lucbv** [comment]
@mayrmt yes, it probably will... for some reason the auto-merge algorithm did not really pick what we would have liked...
In any case I will remove that piece of code again in my next PR on performance related issues.

**Pull request merged**